orangecontrib/spectroscopy/data.py

```diff
@@ -1014,16 +1024,118 @@ def read_v2(self):
         domain = Orange.data.Domain([], None, metas=metas)
         meta_data = Table.from_numpy(domain, X=np.zeros((len(M), 0)),
                                      metas=Meta_data)
+                            
         return waveN, M, meta_data
 
+    def read_v3(self):
+
+        file_channel = str(self.filename.split(' ')[-2]).strip()
+        folder_file = str(self.filename.split(file_channel)[-2]).strip()
+
+        if 'P' in file_channel:
+            self.channel_p = file_channel
+            self.channel_a = file_channel.replace('P', 'A')
+            file_gsf_p = self.filename
+            file_gsf_a = self.filename.replace('P raw.gsf', 'A raw.gsf')
+            file_html = folder_file + '.html'
+        elif 'A' in file_channel:
+            self.channel_a = file_channel
+            self.channel_p = file_channel.replace('A', 'P')
+            file_gsf_a = self.filename
+            file_gsf_p = self.filename.replace('A raw.gsf', 'P raw.gsf')
+            file_html = folder_file + '.html'
+
+
+        data_gsf_a = self._gsf_reader(file_gsf_a)
+        data_gsf_p = self._gsf_reader(file_gsf_p)
+        info = self._html_reader(file_html)
+
+        final_data, parameters, final_metas = self._format_file(data_gsf_a, data_gsf_p, info)
+
+
+        metas = [Orange.data.ContinuousVariable.make("run"),
+                Orange.data.ContinuousVariable.make("row"),
+                Orange.data.ContinuousVariable.make("column"),
+                Orange.data.StringVariable.make("channel")]
+
+        domain = Orange.data.Domain([], None, metas=metas)
+        meta_data = Table.from_numpy(domain, X=np.zeros((len(final_data), 0)),
+                                     metas=np.asarray(final_metas, dtype=object))
+        
+        meta_data.attributes = parameters
+
+        depth = np.arange(0, int(parameters['Pixel Area (X, Y, Z)'][3]))
+
+        return depth, final_data, meta_data
+
+    def _format_file(self, gsf_a, gsf_p, parameters):
+        
+        info = {}
+        for row in range(len(parameters)):
+            info.update({parameters[row,0].replace(':', ''): parameters[row,1:]})
+
+        averaging = int(info['Averaging'][1])
+        px_x = int(info['Pixel Area (X, Y, Z)'][1]) # 10
+        px_y = int(info['Pixel Area (X, Y, Z)'][2]) # 10
+        px_z = number_of_points = int(info['Pixel Area (X, Y, Z)'][3]) # 1024
+        description = info['Description'][1]
+
+        data_complete = []
+        final_metas = []
+        for y in range(0, px_y):
```

Comment:
Just a note... This code is fine for now, but there are many opportunities for better memory and processing efficiency by using numpy matrices in a better way. It is probably possible to avoid most of the for loops. Don't work on this now, we can always improve it later.